File: leetcode-submissions/1405_Longest_Happy_String.ts
function longestDiverseString(a: number, b: number, c: number): string {
    const pq:[number,string][]=[];
    if(a>0){
        pq.push([a,"a"]);
    }
    if(b>0){
pq.push([b,"b"]);

    }
    if(c>0){
pq.push([c,"c"]);
    }
    let res:string="";
    while(pq.length>0){
        pq.sort((a,b)=>b[0]-a[0]);
        let [count,character]=pq.shift();
        
        //after this we need to check whether the string contains consective characters or not 
        if(res.length>=2 && res.charAt(res.length-1)===character && res.charAt(res.length-2)===character){
            if(!pq.length){
                break;
            }
            //otherwise we will need to extract the second largets character count character and add it to the string 
            let [c,ch]=pq.shift();
            res+=ch;
            c--;

            if(c>0){
                pq.push([c,ch]);
            }
        }else{
            res+=character;
            count--;
        }
        if(count>0){
            pq.push([count,character]);
        }
return  res;
    }
};

File: leetcode-submissions/1405_Longest_Happy_String.cpp
class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        priority_queue<pair<int,char>>pq;
        if(a>0){
            pq.push({a,'a'});
        }
        if(b>0){
            pq.push({b,'b'});
        }
        if(c>0){
             pq.push({c,'c'});
        }
        string res="";
        while(not pq.empty()){
            auto [count ,character]=pq.top();
            pq.pop();
            //after removing this we need to keep track 
            int n=res.size();
            if(n>=2 and res[n-1]==character and res[n-2]==character){
                if(pq.empty()){
                    break;
                }
                auto [c,ch]=pq.top();
                pq.pop();
                c--;
                res.push_back(ch);
                if(c>0){
                    pq.push({c,ch});
                }
            }else{
                res.push_back(character);
                count--;
                // pq.push({count,character});
            }
            if(count>0){
                pq.push({count,character});
            }

        }
        return res;
    }
};